libs/server/ArgSlice/ArgSliceUtils.cs

```diff
@@ -15,5 +17,139 @@ public static class ArgSliceUtils
+            // Get the last character to know if this is a quoted context.
+            var type = slice.Span[slice.Length - 1];
+
+            // Nothing to do if it's not.
+            if (!AsciiUtils.IsQuoteChar(type))
+            {
+                return slice;
+            }
+
+            // It's a quoted context, so we need to check for escapes.
+
+            // To short for escaping
+            if (slice.Length <= 3)
+            {
+                return new ArgSlice(slice.ptr + 1, slice.Length - 2);
+            }
+
+            // How many bytes do we need to shift thanks to quoting and escaping.
+            var shift = 0;
+            // start offset
+            var start = 0;
+            // Are we in a quoting context?
+            var qStart = false;
+
+            // Optimize command case by changing our start point instead of shifting the entire string.
+            if (slice.Span[0] == type)
+            {
+                qStart = true;
+                start = 1;
+            }
+
+            for (var j = start; j < slice.Span.Length - shift - 1; ++j)
+            {
+                if (!qStart)
+                {
+                    if (slice.Span[j + shift] == type)
+                    {
+                        qStart = true;
+                        shift++;
+                        slice.Span[j] = slice.Span[j + 1];
+                    }
+                    continue;
+                }
+
+                if (slice.Span[j + shift] != '\\')
+                {
+                    // If we shifted earlier, we need to shift the rest too.
+                    if (shift > 0)
+                        slice.Span[j] = slice.Span[j + shift];
+                    continue;
+                }
+
+                // ' context recognize only this particular sequence
+                if (type == '\'')
+                {
+                    if (slice.Span[j + shift + 1] == '\'')
+                        shift++;
+                    if (shift > 0)
+                        slice.Span[j] = slice.Span[j + shift];
+                    continue;
+                }
+
+                // Process escapes
+                shift++;
+                var c = acceptUppercaseEscapes ?
+                            (char)AsciiUtils.ToLower(slice.Span[j + shift]) :
+                            (char)slice.Span[j + shift];
+
+                switch (c)
+                {
+                    case 'a':
+                        c = '\a';
+                        break;
+                    case 'b':
+                        c = '\b';
+                        break;
+                    case 'n':
+                        c = '\n';
+                        break;
+                    case 'r':
+                        c = '\r';
+                        break;
+                    case 't':
+                        c = '\t';
+                        break;
+                    case 'x':
+                        if (j + shift + 2 < slice.Span.Length)
```

Comment:
I don't love the exception and string allocation here.  I think just write some code to parse the hex directly, or use a lookup table.